core/unix/signal_linux_aarch64.c

```diff
@@ -77,30 +179,107 @@ sigcontext_to_mcontext_simd(priv_mcontext_t *mc, sig_full_cxt_t *sc_full)
     mc->fpsr = fpc->fpsr;
     mc->fpcr = fpc->fpcr;
     ASSERT((sizeof(mc->simd->q) * MCXT_NUM_SIMD_SVE_SLOTS) == sizeof(fpc->vregs));
-    memcpy(&mc->simd, &fpc->vregs, sizeof(mc->simd));
-    /* TODO i#5365: memcpy(&mc->simd->u32,...)
-     * See also sve_context in core/unix/include/sigcontext.h.
-     */
+    int i;
+    for (i = 0; i < MCXT_NUM_SIMD_SVE_SLOTS; i++) {
+        memcpy(&mc->simd[i].q, &fpc->vregs[i], sizeof(mc->simd->q));
+    }
+
+    if (proc_has_feature(FEATURE_SVE)) {
+        size_t offset = sizeof(struct fpsimd_context);
+        /* fpsimd_context is always the first section. After that the esr_context,
+         * extra_context and sve_context sections can be in any order.
+         */
+        struct _aarch64_ctx *next_head =
+            (struct _aarch64_ctx *)(sc_full->sc->__reserved + offset);
+        while (next_head->magic != 0) {
+            ASSERT(next_head->magic == ESR_MAGIC || next_head->magic == SVE_MAGIC ||
+                   next_head->magic == EXTRA_MAGIC);
+            switch (next_head->magic) {
+            case ESR_MAGIC: break;
+            case EXTRA_MAGIC: break;
+            case SVE_MAGIC: {
+                const struct sve_context *sve = (struct sve_context *)(next_head);
+                ASSERT(sve->vl == proc_get_vector_length_bytes());
+                const unsigned int vq = sve_vq_from_vl(sve->vl);
+                if (sve->head.size != sizeof(struct sve_context)) {
+                    for (i = 0; i < MCXT_NUM_SIMD_SVE_SLOTS; i++) {
+                        /* SVE specifies that AArch64's SIMD&FP registers
+                         * (V0-V31) which hold FP scalars and NEON 128-bit
+                         * vectors overlay the bottom 128 bits of the SVE
+                         * registers (Z0-Z31). For backward compatibility
+                         * reasons, bits 0->127 of Z0-Z31 are always restored
+                         * from the corresponding members of fpsimd_context's
+                         * vregs and not from sve_context.
+                         */
+                        memcpy(&mc->simd[i].u32, (byte *)sve + SVE_SIG_ZREG_OFFSET(vq, i),
+                               sve->vl);
+                        memcpy(&mc->simd[i].q, &fpc->vregs[i], sizeof(mc->simd->q));
+                    }
+                    for (i = 0; i < MCXT_NUM_SVEP_SLOTS; i++) {
+                        memcpy(&mc->svep[i].u32, (byte *)sve + SVE_SIG_PREG_OFFSET(vq, i),
+                               sve->vl);
+                    }
+                    memcpy(&mc->ffr, (byte *)sve + SVE_SIG_FFR_OFFSET(vq), sve->vl);
```

Comment:
Were you able to test that these memcpys are doing the right thing?  It is hard to verify as a reviewer.

Reply:
Yes, both `sigcontext_to_mcontext_simd()` and `mcontext_to_sigcontext_simd()` have been tested with scatter/gather and signal tests by dumping out the full context using `dump_sigcontext()` and inspecting the log files. A new function, `dump_ucontext()` implemented in `suite/tests/tools.c` was also used to check context state when running first-fault load tests, which you will see being called in follow-up scatter/gather PRs which depend on this one.